File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys
# input = sys.stdin.readline

input_data = [int(line) for line in sys.stdin]
n, array = input_data[0], input_data[1:]

def solution(n, array):
    answer = []
    stack = []
    cur_num, cur_idx = 1, 0
    
    while cur_idx < n:
        if len(stack) > 0 and stack[-1] == array[cur_idx]: 
            answer.append("-")
            stack.pop()
            cur_idx += 1
        else:
            stack.append(cur_num)
            answer.append("+")
            cur_num += 1
        if len(stack) > 0 and array[cur_idx] < stack[-1]:
            return "NO"
    return "\n".join(answer)


print(solution(n, array))

File: 백준/Silver/1874. 스택 수열/스택 수열.js
const input = require("fs")
  .readFileSync(0)
  .toString()
  .trim()
  .split("\n")
  .map((v) => v.trim())
  .map(Number);
const [N, array] = [input[0], input.slice(1)];

function solution() {
  const answer = [];
  const stack = [];
  let [curNum, curIdx] = [1, 0];

  while (curIdx < N) {
    if (stack[stack.length - 1] === array[curIdx]) {
      answer.push("-");
      stack.pop();
      curIdx++;
    } else {
      stack.push(curNum);
      answer.push("+");
      curNum++;
    }
    if (stack.length > 0 && array[curIdx] < stack[stack.length - 1]) {
      return "NO";
    }
  }
  return answer.join("\n");
}

console.log(solution());
